src/algebra/module/torsion.lean

```diff
@@ -0,0 +1,113 @@
+
+* The notions are defined for a `comm_ring R` and a `module R M`. We further require `R` to be a
+  domain when talking about `torsion' R M` (otherwise it may not be a submodule), and for `M` to be
+  an `add_comm_group` when it's needed.
+* The letters `a`, `b`, ... are used for scalars (in `R`), while `x`, `y`, ... are used for vectors
+  (in `M`).
+
+## Tags
+
+Torsion, submodule, module, quotient
+-/
+
+
+section defs
+variables (R M : Type*) [comm_ring R] [add_comm_monoid M] [module R M] (a : R)
+
+/-- The `a`-torsion submodule, for `a` in `R` -/
+def torsion : submodule R M := (distrib_mul_action.to_linear_map _ _ a).ker
+
+/-- The torsion submodule, only defined when `R` is a domain. -/
+def torsion' [is_domain R] : submodule R M :=
+{ carrier := { x | ∃ a : R, a • x = 0 ∧ a ≠ 0 },
+  zero_mem' := ⟨1, smul_zero _, one_ne_zero⟩,
+  add_mem' := λ x y ⟨a, hx, ha⟩ ⟨b, hy, hb⟩,
+    ⟨b * a,
+      by rw [smul_add, ← smul_smul, mul_comm, ← smul_smul, hx, hy, smul_zero, smul_zero, add_zero],
+      mul_ne_zero hb ha⟩,
+  smul_mem' := λ a x ⟨b, h, hb⟩, ⟨b, by rw [smul_smul, mul_comm, ← smul_smul, h, smul_zero], hb⟩ }
+end defs
+
+section
+variables {R M : Type*} [comm_ring R] [add_comm_monoid M] [module R M] (a : R)
+
+@[simp] lemma smul_torsion (x : torsion R M a) : a • x = 0 := subtype.ext x.prop
+
+/-- A module is torsion-free (`no_zero_smul_divisors`) iff its torsion submodule is trivial. -/
+lemma no_zero_smul_divisors_iff_torsion_bot [is_domain R] :
+  no_zero_smul_divisors R M ↔ torsion' R M = ⊥ :=
+begin
+  split; intro h,
+  { haveI : no_zero_smul_divisors R M := h,
+    ext, split; intro hx,
+    { cases hx with a hax,
+      cases eq_zero_or_eq_zero_of_smul_eq_zero (hax.left) with h0 h0,
+      { exfalso, exact hax.right h0 }, { exact h0 } },
+    { have hx : x = 0 := hx, rw hx, exact (torsion' R M).zero_mem } },
+  { exact { eq_zero_or_eq_zero_of_smul_eq_zero := λ a x hax, begin
+      by_cases ha : a = 0,
+      { left, exact ha },
+      { right, rw [← submodule.mem_bot _, ← h], exact ⟨a, hax, ha⟩ }
+    end } }
+end
+end
+
+section quotient
+open ideal.quotient
+open submodule.quotient
+variables {R M : Type*} [comm_ring R] [add_comm_group M] [module R M] (a : R)
+
+instance : has_scalar (R ⧸ ideal.span ({a} : set R)) (torsion R M a) :=
+  { smul := λ b x, quotient.lift_on' b (• x) $ λ b₁ b₂ (h : b₁ - b₂ ∈ _),
+    show b₁ • x = b₂ • x, begin
+    obtain ⟨c, h⟩ := ideal.mem_span_singleton'.mp h,
+    rw [← sub_eq_zero, ← sub_smul, ←h, mul_smul, smul_torsion, smul_zero],
+  end }
+
+@[simp] lemma mk_smul' (r : R) (x : torsion R M a) : mk (ideal.span ({a} : set R)) r • x = r • x :=
+rfl
+
+/-- The `a`-torsion submodule as a `(R ⧸ R·a)`-module. -/
+instance : module (R ⧸ ideal.span ({a} : set R)) (torsion R M a) :=
+function.surjective.module_left (mk _) (mk_surjective _) (mk_smul' _)
+
+instance : is_scalar_tower R (R ⧸ ideal.span ({a} : set R)) (torsion R M a) :=
+{ smul_assoc := λ b d x, by { rw [← mk_smul', smul_smul], refl } }
```

Comment:
This should generalize to something like
```suggestion
instance [is_scalar_tower S R R] : is_scalar_tower S (R ⧸ ideal.span ({a} : set R)) (torsion R M a) :=
{ smul_assoc := λ b d x, by { rw [← mk_smul', smul_assoc], refl } }
```
(untested)

Reply:
in fact you want an
```suggestion
instance [has_scalar A B] [has_scalar B C] [has_one B] : has_scalar A C := { smul := \fun a c, (a \bu (1:B)) \bu c }
```
and from there you get an [is_scalar_tower A B C] if the smul are compatible with the 1 on B (in fact you also need a has_mul B to get it work)
and also the equality of the two [has_scalar A D] instances you get from [has_scalar A B] [has_scalar B C] [has_scalar C D] [has_one B] [has_one C] (from building either [has_scalar A C] or [has_scalar B D] first) under good conditions

Reply:
Here's the working version:
```lean
instance [has_scalar S R] [has_scalar S M] [is_scalar_tower S R M] [is_scalar_tower S R R] :
  is_scalar_tower S (R ⧸ ideal.span ({a} : set R)) (torsion R M a) :=
{ smul_assoc := λ b d x, quotient.induction_on' d $ λ d, (smul_assoc b d x : _) }
```

Reply:
Thanks for the clarification